Step2 리뷰요청드립니다.

lcjltj opened the pull request:
안녕하세요! 리뷰어님.

1단계에서 피드백 주신 부분 반영및 최단 경로 조회 로직 및 인수테스트 작성해 보았습니다.!

잘부탁드립니다.

lcjltj committed:
refactor: step1 feedback 반영

- Section / Distance 도메인 테스트 케이스 추가
- Service @Transaction readonly 적용
- Line color column nullable 추가

lcjltj committed:
feat: Jgraph 를 통한 최단 경로 탐색 기능 추가 외 테스트 케이스 작성

- getShortestPath() 다익스트라 최단 경로 구하는 메소드 추가

lcjltj committed:
feat: 서비스 로직 추가 / 최단 경로 인수 테스트 작성

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -13,6 +14,7 @@ public class Line extends BaseEntity {
     private Long id;
     @Column(unique = true)
     private String name;
+    @Column(nullable = false)
```
피드백 반영 👍

ohtaeg commented on src/main/java/nextstep/subway/path/advice/PathAdvice.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice(assignableTypes = {PathController.class})
```
`@ControllerAdvice` 👍 
아실수도 있지만 `@RestControllerAdvice`랑 어떤 차이가 있는지 알아보시는 것도 좋을 것 같아요 😄

lcjltj replied:
@RestControllerAdvice를 사용했어야 했는데 습관이란게 무섭네요 ! ㅜ-ㅜ
@RestControllerAdvice가  @Controller와 @RestController  차이처럼 @ResponseBody 로 return으로 할수 있게 했어야하는데 감사합니다.!

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.List;
+
+@Service
+@Transactional
+public class PathService {
```
최단 경로를 조회하는 서비스를 만들어주셨네요 👍

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,38 @@
+@Service
+@Transactional
+public class PathService {
+    private final StationService stationService;
+    private final LineService lineService;
+
+    public PathService(StationService stationService, LineService lineService) {
+        this.stationService = stationService;
+        this.lineService = lineService;
+    }
```
이 부분은 같이 고민해보면 좋을 것 같아요 :)
저는 개인적으로 많은 의존성을 가질 확률이 높은 Service 레이어를 의존하는게 좋을지 고민이 되는 부분이네요!
그렇다고 예외 처리나 조회에 있어서 일관성 있는 방법을 위해 Service를 의존하는 방향도 틀렸다고 생각하지 않는데요!
트레이드 오프 관점이라 고민이 되는 부분이에요 ㅎㅎ
찬준님은 어떻게 생각하시나요 :) ?

lcjltj replied:
안녕하세요 마침. 어제 강의에서 도메인 서비스/ 응용 서비스 관련한 내용을 말해주셔서 해당 부분에 대해서는 고민이 많네요. 리뷰어 님께서는 위 로직을 작성했을경우 repository를 의존하여 개발을 하는 식으로 하실까요? 
개인적인 생각으로는 도메인 서비스에서 repository 를 사용하지 않고 응용 서비스를 따로 하나 만들어 DB에 접근 하여 받아올 데이터를 받고 PathService(도메인서비스)는 어떠한 의존성도 없이 개발하는게 좋을거 같긴한데.
의견 부탁드립니다!!

ohtaeg replied:
우선 이 부분은 정답이 없는 부분인 점 참고 부탁드립니다 :)
팀 내 협의에 따라 달라질 수 있지만 현재 구조로 논의를 해보자면 조금 어려울 것 같아요!
애그리거트를 어떻게 묶을 것인지 부터 다시 고민해봐야 할 것 같아요!
현재 패키지는 path, line, station 패키지가 분리되어 있기에 이 각 패키지를 각 애그리거트로 한다면 말씀해주신대로
PathService(도메인서비스)는 어떠한 서비스와 의존성 없이 외부로부터 애그리거트를 받는 식으로 할 것 같습니다 :) 

DDD를 떠나서 현재 구조라면 패키지 기준으로 각 모듈로 분리된것이라고 생각을 해보면
모듈간 결합성을 떨어트리기 위해 infra 영역인 Repository를 참조할 것 같아요 :)
실제 실무에서 서비스가 서비스를 참조하는 형태에서는 다양한 부작용들이 있을 수 있다고 생각합니다!
순환참조가 대표적인 예이며, 만약 해당 주입받는 서비스가 의존성이 많은 경우 테스트도 쉽지않구요.

아래 링크도 한번 참고해보시면 좋으실 것 같아요 😄 
<img width="642" alt="image" src="https://user-images.githubusercontent.com/20432323/175458879-f46790f5-24e0-461c-a570-10721a3f89da.png">

https://martinfowler.com/bliki/PresentationDomainDataLayering.html

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,38 @@
+        final Station destination = stationService.findById(destinationStationId);
+        final List<Line> lines = lineService.findAll();
+
+        final PathFinder pathFinder = new PathFinder();
+        final GraphPath<Station, DefaultWeightedEdge> graphPath = pathFinder.getShortestPath(lines, start, destination);
```
이 부분은 같이 고민해보면 좋을 것 같아요!
현재 PathFinder를 도메인 패키지에 위치하신 걸 통해 도메인 모델이라고 짐작이 가는데요!
외부 라이브러리를 이용한 최단 경로 찾는 방법은 언제든지 바뀔 수 있는 점을 고려해보았을 때

만약 다른 최단경로알고리즘 라이브러리를 사용하거나 직접 구현할 경우 알고리즘 교체에 유연한지 고민해보면 좋을 것 같아요 :)
게다가 알고리즘 변경이 되면 도메인 내부에서 변경이 일어나기에 
인터페이스를 선언하고, 실제 구현체는 외부에서 주입받아 외부 라이브러리 구현체에 직접 의존하는 것을 최소화하면 어떨까 싶습니다! 찬준님은 이 부분에 대해서는 어떻게 생각하시나요 :) ?

lcjltj replied:
해당 로직을 개발할때 인터페이스를 두어 의존성 관계를 구현체에 하지 않으려고 했었는데.
그 경계가 너무 어렵네요.  다른 방식으로 하게 될 경우 인터페이스를 추가하면서 분리를 할 생각도 있었는데 강의 전에 미리 예상하고 인터페이스를 분리하는 것도 낭비를 하는거라는 말을 들은게 있어서. 그렇게 하지 않았었네요.
앞으로는 외부 라이브러리를 사용하는 경우에 인터페이스를 따로 두어 의존성을 최소화하도록 하겠습니다.

ohtaeg replied:
의견을 공유해주셔서 감사드립니다 👍 👍 

> 미리 예상하고 인터페이스를 분리하는 것도 낭비

저도 이말에 동감합니다 💯  미리 추상화하는 것 자체가 비용이라고 생각하는데요!
저도 처음은 최단 경로 찾는 객체는 찬준님처럼 했을거에요 :)
리뷰 드리다보니 의존성을 고민하게 되었습니다 😄 
물론 해당 리뷰가 정답이 아니기에 같이 고민해보면 좋을 것 같아 말씀드린 부분이기도 하였어요 👍

ohtaeg commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -0,0 +1,52 @@
+public class PathFinder {
+    private WeightedMultigraph<Station, DefaultWeightedEdge> graph;
+
+    public GraphPath<Station, DefaultWeightedEdge> getShortestPath(final List<Line> lines, final Station start, final Station destination) {
```
외부 라이브러리의 GraphPath 의존성을 PathFinder를 사용하는 서비스 레이어로 반환하는 부분에 대해 고민해보면 좋을 것 같아요! GraphPath 대신 GraphPath.getVertexList()의 결과인 List\<Station>와 가중치를 반환하되
두 값을 가지는 PathResponse나 VO성 객체를 반환해도 좋을 것 같은데 어떻게 생각하시나요 :) ?

lcjltj replied:
return type을 PathResponse로 하게 되면 PathResponse에 의존하게 되어 해당 로직에서 하지 않고,  서비스 부분에서만 하게 했었습니다.  로직을 보니, VO같은 객체로 해당 값을 반환 하여,  해당 데이터를 사용하는 방식이 좋아 보여, 수정 하였습니다.
바로 PathResponse 로 return하는것과 VO를 생성해서 하는것에 대한 리뷰어님의 생각은 어떠신가요?

ohtaeg replied:
> PathResponse에 의존하게 되어 해당 로직에서 하지 않고, 서비스 부분에서만 하게 했었습니다.

말씀해주신대로 VO가 적절해보여요 👍 👍

ohtaeg commented on src/test/java/nextstep/subway/line/domain/DistanceTest.java:
```diff
@@ -0,0 +1,51 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class DistanceTest {
```
👍 👍

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -1,9 +1,70 @@
+import static org.assertj.core.api.InstanceOfAssertFactories.ARRAY;
 
 @DisplayName("지하철 경로 조회")
 public class PathAcceptanceTest extends AcceptanceTest {
```
최단 경로를 구하는 인수 테스트 및 도메인 테스트도 작성을 잘해주셨어요 👍

ohtaeg reviewed (changes requested):
안녕하세요 찬준님! 2단계 미션도 잘해주셨네요 👍 👍 
미션 요구사항을 잘 만족시켜주셔서 merge 하려했으나,
함께 고민하면 좋을 포인트들이 존재하여 Request changes 하였습니다 :)
몇 개의 코멘트를 드렸는데 확인해주시면 감사하겠습니다 🙏

lcjltj committed:
refactor: 피드백 반영 (인터페이스 분리 외

 - PathFinder 인터페이스 분리
 - StationPath VO 생성
 - PathAdvice @ControllerAdvice -> @RestControllerAdivce 변경

ohtaeg commented on src/main/java/nextstep/subway/path/advice/PathAdvice.java:
```diff
@@ -3,10 +3,10 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
 
-@ControllerAdvice(assignableTypes = {PathController.class})
+@RestControllerAdvice(assignableTypes = {PathController.class})
```
앗 반영해주셨네요 👍

ohtaeg commented on src/main/java/nextstep/subway/path/domain/PathFinder.java:
```diff
@@ -1,52 +1,10 @@
-        final DefaultWeightedEdge edge = graph.addEdge(section.getUpStation(), section.getDownStation());
-        graph.setEdgeWeight(edge, section.getDistance());
-    }
+public interface PathFinder {
+    StationPath getShortestPath(final List<Line> lines, final Station start, final Station destination);
```
피드백 반영 👍

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨어요 👍 👍 
주셨던 코멘트에 대해 의견을 드렸는데 확인해주시면 감사하겠습니다 :)
다음 미션도 화이팅하세요 💪 불금 잘 보내시구요 🔥
